dlib/matrix/fft_size.h

```diff
@@ -0,0 +1,97 @@
+#ifndef DLIB_FFT_SIZE_H
+#define DLIB_FFT_SIZE_H
+
+#include <array>
+#include <algorithm>
+#include <numeric>
+#include "../assert.h"
+#include "../hash.h"
+
+namespace dlib
+{
+    class fft_size
```

Comment:
I find this class confusing to reason about.  Like it's not clear what it represents at any moment and what the valid states are when I look at the code that uses it.  I mean, I can follow what's happening, but it's subtle.

What I meant was something like this:
```
class fft_size
{
public:
    explicit fft_size(long x_dim) : x_(x_dim){}
    /*! 
        ensures
            - x() == x_dim
            - y() == 1
    !*/
    fft_size(long x_dim, long y_dim) : x_(x_dim), y_(y_dim) {}
    /*! 
        ensures
            - x() == x_dim
            - y() == y_dim
    !*/

    // get size of dimensions.
    long x() const { return x_; }
    long y() const { return y_; }

    long num_elements() const { return x_*y_; }

    bool operator==(const fft_size& other) const;
    uint32 hash() const;
private:
    
    long x_ = 1;
    long y_ = 1;
};
```

That object is really simple and immutable.  It always represents the same thing.  There isn't optionality about being 1D sometimes and 2D other times.  A 1D FFT just means y==1.  If 3D ffts are needed later someone can add a z (and even a w) dimension.

You could add global helper functions for the higher level stuff you are doing with it.  For example, you might say ```
if (is_1d(some_fft_size))
    // do whatever
```
and that would make the code that uses it more readable.

Reply:
Ok. Honestly i don't mind. If I need to go 3D and beyond at some point, I can always submit another PR. Whatever you think is best.

Reply:
Though ```x_``` and ```y_``` may be confusing... By ```x_``` do you mean the fft size along the column dimension and by ```y_``` do you mean the fft size along the row dimension? In which case, the order of FFTs is ```y_``` then ```x_```.  Lets go with that, but i find this class more confusing. Whereas a simple vector-like class that describes arbitrary dimensions seems more appropriate in the KISS and MKL backends. Indeed the fft sizes are dims[0], dims[1], ..., dims[N-1]. The interpretation of what the dimensions mean seems more appropriately dealt with in ```matrix_fft.h```. But that's my opinion. I'll go with what you want since dlib is your repo at the end of the day.

Reply:
In other words, the ordering of the dimensions is not at all obvious now.

Reply:
And going with the ordering of the alphabet is still not clear because, typically, matrix layout is row major, so ordering goes ```y``` then ```x```.

Reply:
x is the first dimension, y the next.  But yeah, I see what you mean.  `dims[0]` and `dims[1]` is better. The main thing is to make `fft_size` a simple immutable container that always has the same number of integers in it.

Reply:
So what do you suggest? i had ```std::array<long,5>``` do cater for up to 5. But i could have ```std::array<long,2>``` and not allow it to be empty.

Reply:
Unless it simplifies the code that uses it names like `dim0` and `dim1` would be best, since code that does `dim2` is a compile error while `dim[2]` is not.  It doesn't look like there needs to be runtime indexing into these things.  If that's so then `dim0` and `dim1` seem good.

To your question, put 2 longs in there since that's what's actually used right now.  If it needs to be more later it can be changed.

Reply:
My problem with this is that we are simplifying backend code when there is no need to. I understand making the code in ```matrix_fft.h``` as full-proof as possible but i don't see any reason to in the implementations details. The user should have no visibility of this. If we go down the rabbit hole of specialising stuff, we could go much further and replace the ND case for 2D only. Then the 2D plan could have 2 1D plans instead of a std::vector of 1D plans. The real ND plan currently has 1 real 1D plan and 1 (complex) ND plan. We could change that to 1 complex 1D plan and 1 real 1D plan. And so forth... This would be undoing all the hard work of supporting ND FFTs. I know what you suggest is simply refactoring ```fft_size``` object. But it makes no sense to simplify a tiny skin depth layer of the backend. Either simplify everything or leave it general as it is now.

Reply:
The original point of ```fft_size``` was to remove dynamic allocations, which was achieved...

Reply:
For me, ```fft_size``` is just a much simpler version of boost's ```static_vector``` to replace ```std::vector``` with something that doesn't perform dynamic memory allocations. And that's it. It's just a container for FFT dimensions (with a hard upper bound of 5)

Reply:
All of this to say, i find the proposed amendment a bit itchy. It's like, "Why am I restricting all the KISS FFT stuff to 1d and 2d but all the rest of the code seems to support, or is set up to support, ND case. That's weird... Something has gone wrong somewhere along the way...."

Reply:
Yeah, don't undo work in kissfft.  What's in here is just hard for me to follow.  Like it's not clear what exactly is supported and supposed to work and what isn't at first.  Like it's not obvious what kinds of `fft_size` objects are valid and meaningful and which aren't. E.g. Are size 0 ones generally ok?  Setting it up so the full power of kissfft is part of the kissfft interface (so the kissfft functions, the plans, and the `fft_size` object) is preferable though.  So making the number of dimensions in `fft_size` a runtime option and letting it be anything is fine.  But make the interface clear.  Just adding comments that say what these interfaces claim to support would maybe be all that's really required.  Although making `fft_size` immutable would be a good thing regardless.

I guess I'm saying that it's hard to follow sine most of these things in the diff have no comments.  So it's unclear what the exact contracts for some of these functions are.

Reply:
By making ```fft_size``` immutable, do you mean it can't be copyable or moveable, or do you mean all of its member functions must be const ?

Reply:
For example, i can make a function ```remove_ones()``` a member function that isn't const, or i can make a static function that takes a ```const fft_size&``` and returns a ```fft_size```. The latter is just functional style programming. The same applies to a function like ```pop_back()```. I can make it a non-const member function or create a functional style equivalent.

Reply:
I mean all the members are const.

Reply:
Don't make any class static methods.  Just make global methods.

Reply:
Cool